src/main/java/com/exasol/adapter/dynamodb/documentpath/DocumentPathWalker.java

```diff
@@ -0,0 +1,129 @@
+
+    /**
+     * Creates an instance of {@link DocumentPathWalker}.
+     * 
+     * @param pathExpression path to walk
+     */
+    public DocumentPathWalker(final DocumentPathExpression pathExpression) {
+        this.pathExpression = pathExpression;
+    }
+
+    /**
+     * Walks the path defined in constructor through the given document.
+     * 
+     * @param rootNode document to walk through
+     * @return documents attributes described in {@link DocumentPathExpression}.
+     * @throws DocumentPathWalkerException if defined path does not exist in the given document
+     */
+    public List<DocumentNode<VisitorType>> walk(final DocumentNode<VisitorType> rootNode)
+            throws DocumentPathWalkerException {
+        return this.walk(rootNode, 0);
+    }
+
+    private List<DocumentNode<VisitorType>> walk(final DocumentNode<VisitorType> thisNode, final int position)
+            throws DocumentPathWalkerException {
+        if (this.pathExpression.size() <= position) {
+            return List.of(thisNode);
+        }
+        final Function<DocumentNode<VisitorType>, List<DocumentNode<VisitorType>>> converter = getConverterFor(
+                this.pathExpression.getPath().get(position));
+        try {
+            final List<DocumentNode<VisitorType>> nextNodes = converter.apply(thisNode);
+            final ArrayList<DocumentNode<VisitorType>> results = new ArrayList<>();
+            for (final DocumentNode<VisitorType> nextNode : nextNodes) {
+                results.addAll(walk(nextNode, position + 1));
+            }
+            return results;
+        } catch (final InternalPathWalkException exception) {
+            throw addCurrentPathToException(exception, position);
+        }
+    }
+
+    private DocumentPathWalkerException addCurrentPathToException(final InternalPathWalkException exception, final int position)  {
+        final DocumentPathExpression currentPath = this.pathExpression.getSubPath(0, position);
+        final String currentPathString = new DocumentPathToStringConverter().convertToString(currentPath);
+        return new DocumentPathWalkerException(exception.getMessage(), currentPathString);
+    }
+
+    @java.lang.SuppressWarnings("squid:S119") // VisitorType does not fit naming conventions.
+    private Function<DocumentNode<VisitorType>, List<DocumentNode<VisitorType>>> getConverterFor(
+            final PathSegment pathSegment) {
+        final WalkVisitor<VisitorType> visitor = new WalkVisitor<>();
+        pathSegment.accept(visitor);
+        return visitor.converter;
+    }
+
+    @java.lang.SuppressWarnings("squid:S119") // VisitorType does not fit naming conventions.
+    private static class WalkVisitor<VisitorType> implements PathSegmentVisitor {
+        Function<DocumentNode<VisitorType>, List<DocumentNode<VisitorType>>> converter;
+
+        @Override
+        public void visit(final ObjectLookupPathSegment objectLookupPathSegment) {
+            this.converter = thisNode -> {
```

Comment:
What is this `->` notation? I did not expect this in Java codebase :))

From my understanding, converter is a function that takes node as an argument, and the body in inside the curly braces, right?